packages/gatsby/cache-dir/loader.js

```diff
@@ -353,9 +372,20 @@ const createComponentUrls = componentChunkName =>
 export class ProdLoader extends BaseLoader {
   constructor(asyncRequires, matchPaths) {
     const loadComponent = chunkName =>
-      asyncRequires.components[chunkName]().then(preferDefault)
+      asyncRequires.components[chunkName]()
+        .then(preferDefault)
+        .catch(
+          () =>
+            // failed to load component but not an unhandled exception
+            // because we handle case of component resolving to null
+            null
+        )
 
-    super(loadComponent, matchPaths)
+    // check if component is available in chunks dictionary
+    const pageDataValidator = pageData =>
+      !!asyncRequires.components[pageData.componentChunkName]
```

Comment:
Controversial:
What this does it adds additional validation to page-data. We currently check if file is event a json file and also if it contains `path` field.

This adds check to see if currently loaded bundle have listed page template inside (we need some check for it, because otherwise we will crash with `asyncRequires.components[pageData.componentChunkName] is not a function` error while we try to `loadComponent`.

It's marked as controversial here, because of the flow that happens when page-data is considered invalid - we load 404 (in browser).

There can be case been made that we instead should try to navigate to this page using regular browser navigation: loaded webpack bundle might be already outdated (there was new deploy happening), tho it's not clear how to best achieve that. Navigating using regular browser navigation (not clientside one) would automatically reload bundles (if html site - either existing one or 404 one have updated script resources).